def helpMessage() {
    log.info """
    =======================================================
                                              ,--./,-.
              ___     __   __   __   ___     /,-._.--~\'
        |\\ | |__  __ /  ` /  \\ |__) |__         }  {
        | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,
                                              `._,._,\'

     nf-core/rnaseq : RNA-Seq Best Practice v${workflow.manifest.version}
    =======================================================

    Usage:

    The typical command for running the pipeline is as follows:

    nextflow run nf-core/rnaseq --reads '*_R{1,2}.fastq.gz' --genome GRCh37 -profile uppmax

    Mandatory arguments:
      --reads                       Path to input data (must be surrounded with quotes)
      --readPathFile                Tab-seperated file with sample names and path to the fastq files. (Used if --reads not provided.)
      -profile                      Configuration profile to use. uppmax / uppmax_modules / hebbe / docker / aws

    Additional quantification options:
      --skip_alignment              Skip alignment with STAR or HISAT2 (deafault: false)
      --run_salmon                  Runs transcript expression quantification (Salmon)
      --run_txrevise                Runs txrevise quantification (Salmon with custom reference transciptome)
      --run_leafcutter              Runs alternative splicing quantification  (LeafCutter)
      --run_exon_quant              Runs exon quantification (DEXseq)

    Options:
      --genome                      Name of iGenomes reference
      --singleEnd                   Specifies that the input is single end reads
    Strandedness:
      --forward_stranded            The library is forward stranded
      --reverse_stranded            The library is reverse stranded
      --unstranded                  The default behaviour

    References                      If not specified in the configuration file or you wish to overwrite any of the references.
      --star_index                  Path to STAR index
      --hisat2_index                Path to HiSAT2 index
      --fasta                       Path to Fasta reference
      --gtf                         Path to GTF file
      --gff                         Path to GFF3 file
      --bed12                       Path to bed12 file
      --saveReference               Save the generated reference files the the Results directory.
      --saveTrimmed                 Save trimmed FastQ file intermediates
      --saveAlignedIntermediates    Save the BAM files from the Aligment step  - not done by default

    Trimming options
      --clip_r1 [int]               Instructs Trim Galore to remove bp from the 5' end of read 1 (or single-end reads)
      --clip_r2 [int]               Instructs Trim Galore to remove bp from the 5' end of read 2 (paired-end reads only)
      --three_prime_clip_r1 [int]   Instructs Trim Galore to remove bp from the 3' end of read 1 AFTER adapter/quality trimming has been performed
      --three_prime_clip_r2 [int]   Instructs Trim Galore to re move bp from the 3' end of read 2 AFTER adapter/quality trimming has been performed

    Presets:
      --pico                        Sets trimming and standedness settings for the SMARTer Stranded Total RNA-Seq Kit - Pico Input kit. Equivalent to: --forward_stranded --clip_r1 3 --three_prime_clip_r2 3
      --fcExtraAttributes           Define which extra parameters should also be included in featureCounts (default: gene_names)

    Other options:
      --outdir                      The output directory where the results will be saved
      -w/--work-dir                 The temporary directory where intermediate data will be saved
      --email                       Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits
      --sampleLevel                 Used to turn of the edgeR MDS and heatmap. Set automatically when running on fewer than 3 samples
      --clusterOptions              Extra SLURM options, used in conjunction with Uppmax.config
      --maxMultiqcEmailFileSize     Theshold size for MultiQC report to be attached in notification email. If file generated by pipeline exceeds the threshold, it will not be attached (Default: 25MB)
      -name                         Name for the pipeline run. If not specified, Nextflow will automatically generate a random mnemonic.
      --seqCenter                   Add sequencing center in @RG line of output BAM header

    QC options:
      --skip_qc                     Skip all QC steps apart from MultiQC
      --skip_fastqc                 Skip FastQC
      --skip_rseqc                  Skip RSeQC
      --skip_genebody_coverage      Skip calculating genebody coverage
      --skip_preseq                 Skip Preseq
      --skip_dupradar               Skip dupRadar (and Picard MarkDups)
      --skip_edger                  Skip edgeR MDS plot and heatmap
      --skip_multiqc                Skip MultiQC

    AWSBatch options:
      --awsqueue                    The AWSBatch JobQueue that needs to be set when running on AWSBatch
      --awsregion                   The AWS Region for your AWS Batch job to run on
    """.stripIndent()
}

def summaryMessage() {
  custom_runName = params.name

  if( !(workflow.runName ==~ /[a-z]+_[a-z]+/) ){
    custom_runName = workflow.runName
  }

  log.info """=======================================================
                                            ,--./,-.
            ___     __   __   __   ___     /,-._.--~\'
      |\\ | |__  __ /  ` /  \\ |__) |__         }  {
      | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,
                                            `._,._,\'

  nf-core/rnaseq : RNA-Seq Best Practice v${workflow.manifest.version}
  ======================================================="""
  def summary = [:]
  summary['Run Name']     = custom_runName ?: workflow.runName
  summary['Reads']        = params.reads
  summary['ReadPathsFile']        = params.readPathsFile
  summary['Data Type']    = params.singleEnd ? 'Single-End' : 'Paired-End'
  summary['Genome']       = params.genome
  if( params.pico ) summary['Library Prep'] = "SMARTer Stranded Total RNA-Seq Kit - Pico Input"
  summary['Strandedness'] = ( params.unstranded ? 'None' : params.forward_stranded ? 'Forward' : params.reverse_stranded ? 'Reverse' : 'None' )
  summary['Trim R1'] = params.clip_r1
  summary['Trim R2'] = params.clip_r2
  summary["Trim 3' R1"] = params.three_prime_clip_r1
  summary["Trim 3' R2"] = params.three_prime_clip_r2
  if(params.aligner == 'star'){
      summary['Aligner'] = "STAR"
      if(params.star_index)          summary['STAR Index']   = params.star_index
      else if(params.fasta)          summary['Fasta Ref']    = params.fasta
  } else if(params.aligner == 'hisat2') {
      summary['Aligner'] = "HISAT2"
      if(params.hisat2_index)        summary['HISAT2 Index'] = params.hisat2_index
      else if(params.fasta)          summary['Fasta Ref']    = params.fasta
      if(params.splicesites)         summary['Splice Sites'] = params.splicesites
  }
  if(params.gtf)                 summary['GTF Annotation']  = params.gtf
  if(params.gff)                 summary['GFF3 Annotation']  = params.gff
  if(params.bed12)               summary['BED Annotation']  = params.bed12
  summary['Save Reference'] = params.saveReference ? 'Yes' : 'No'
  summary['Save Trimmed']   = params.saveTrimmed ? 'Yes' : 'No'
  summary['Save Intermeds'] = params.saveAlignedIntermediates ? 'Yes' : 'No'
  summary['Save Indv Quants']  = params.saveIndividualQuants ? 'Yes' : 'No'
  summary['Max Memory']     = params.max_memory
  summary['Max CPUs']       = params.max_cpus
  summary['Max Time']       = params.max_time
  summary['Output dir']     = params.outdir
  summary['Skip alignment'] = params.skip_alignment
  summary['Run salmon']     = params.run_salmon
  summary['Run exon quant'] = params.run_exon_quant
  summary['Run leafcutter'] = params.run_leafcutter
  summary['Run txrevise']   = params.run_txrevise
  summary['Working dir']    = workflow.workDir
  summary['Container']      = workflow.container
  if(workflow.revision) summary['Pipeline Release'] = workflow.revision
  summary['Current home']   = "$HOME"
  summary['Current user']   = "$USER"
  summary['Current path']   = "$PWD"
  summary['Script dir']     = workflow.projectDir
  summary['Config Profile'] = workflow.profile
  if(params.project) summary['UPPMAX Project'] = params.project
  if(params.email) {
      summary['E-mail Address'] = params.email
      summary['MultiQC maxsize'] = params.maxMultiqcEmailFileSize
  }
  log.info summary.collect { k,v -> "${k.padRight(15)}: $v" }.join("\n")
  log.info "========================================="


  // Show a big error message if we're running on the base config and an uppmax cluster
  if( workflow.profile == 'standard'){
      if ( "hostname".execute().text.contains('.uppmax.uu.se') ) {
          log.error "====================================================\n" +
                    "  WARNING! You are running with the default 'standard'\n" +
                    "  pipeline config profile, which runs on the head node\n" +
                    "  and assumes all software is on the PATH.\n" +
                    "  ALL JOBS ARE RUNNING LOCALLY and stuff will probably break.\n" +
                    "  Please use `-profile uppmax` to run on UPPMAX clusters.\n" +
                    "============================================================"
      }
  }
}